src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -388,16 +380,23 @@ internal static void RunForThreadPoolUserWorkItemWithDefaultContext<TState>(Acti
 
             callback.Invoke(state);
 
-            Thread currentThread = Thread.CurrentThread;
+            // ThreadPoolWorkQueue.Dispatch will handle notifications and reset EC and SyncCtx back to default
+        }
+
+        internal static void ResetThreadPoolThread(Thread currentThread)
+        {
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
             if (currentExecutionCtx != null && currentExecutionCtx.HasChangeNotifications)
             {
                 // Restore to Default before Notifications, as the change can be observed in the handler
+                currentThread.SynchronizationContext = null;
                 currentThread.ExecutionContext = null;
                 OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);
             }
 
-            // ThreadPoolWorkQueue.Dispatch will reset EC and SyncCtx back to default
+            // Reset back to defaults in case the Change handler changed the contexts
```

Comment:
This is the main reset, maybe the comment could be tweaked or it could be written a bit differently:

```c#
// Reset to defaults
currentThread.ExecutionContext = null;
currentThread.SynchronizationContext = null;

if (currentExecutionCtx != null && currentExecutionCtx.HasChangeNotifications)
{
    OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);

    // Reset to defaults again without change notifications in case the Change handler changed the contexts
    currentThread.ExecutionContext = null;
    currentThread.SynchronizationContext = null;
}
```

Reply:
I like that

Reply:
At this point default context UWI can just run rather than using EC.Run, which makes it the same as unsafe run, so unsafe can use the `QueueUserWorkItemCallbackDefaultContext` version?

Reply:
Since its not really doing anything
```csharp
internal static void RunForThreadPoolUserWorkItemWithDefaultContext<TState>(Action<TState> callback, in TState state)
{
    CheckThreadPoolAndContextsAreDefault();

    callback.Invoke(state);

    // ThreadPoolWorkQueue.Dispatch will handle notifications and reset EC and SyncCtx back to default
}
```

Reply:
Yea true any of the default EC cases could use `QueueUserWorkItemCallbackDefaultContext` instead

Reply:
So would be the last change now? https://github.com/dotnet/coreclr/pull/20308/commits/b412b4f453a05f7e21ef29a9100446ce743705cc